### PR TITLE
[Neuron] Add custom_ops for neuron backend

### PR DESCRIPTION
**Pull request by liangfu**
As part of the effort in supporting vLLM V1 architecture for neuron backend (https://github.com/vllm-project/vllm/issues/11152), this PR intent to support activation, layernorm, rotary_embedding and logits_processor as a variant of existing modules that derive from CustomOp class.

These changes are tested in individual test cases that are ported from `tests/kernels` directory to be neuron backend specific.

Co-authored-by: George Novack (@gnovack)
Co-authored-by: Aoyu Zhang (@AoyuQC)

### CODE REVIEW

File: tests/neuron/test_activation.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    current_platform.seed_everything(0)
+    torch.set_default_device("cpu")
+    x = torch.randn(num_tokens, 2 * d, dtype=dtype).to(device=device)
+    if activation == "silu_and_mul":
+        layer = SiluAndMul()
+
+        def _silu_and_mul(x: torch.Tensor) -> torch.Tensor:
+            assert x.is_cpu, "reference input is expected be executed on cpu."
+            d = x.shape[-1] // 2
+            return F.silu(x[..., :d]) * x[..., d:]
```
Comment (gnovack) [inline tests/neuron/test_activation.py:36]: can we use `layer.forward_native` here instead of redefining the operation?

File: tests/neuron/test_rotary_embedding.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+import pytest
+import torch
+
+from vllm.model_executor.layers.rotary_embedding import RotaryEmbedding
+from vllm.platforms import current_platform
+
+
+@pytest.mark.parametrize(
+    "max_position,is_neox_style,rotary_dim,head_size,seq_len", [
```
Comment (gnovack) [inline tests/neuron/test_rotary_embedding.py:14]: We should add a test case where `rotary_dim != head_size`

File: vllm/model_executor/layers/activation.py [hunk 3]
```diff
@@ -89,6 +89,19 @@ def forward_xpu(self, x: torch.Tensor) -> torch.Tensor:
...
+        d = x.shape[-1] // 2
+        if len(x.shape) == 3:
+            s = x[:, :, :d] * torch.nn.functional.sigmoid(x[:, :, :d])
+            return s * x[:, :, d:]
+        elif len(x.shape) == 2:
+            s = x[:, :d] * torch.nn.functional.sigmoid(x[:, :d])
+            return s * x[:, d:]
+        else:
+            raise NotImplementedError(
+                "Expected input to have either 3 or 2 dims")
```
Comment (gnovack) [inline vllm/model_executor/layers/activation.py:103]: We can probably find a cleaner way to implement this. Initially did it this way because the ellipsis (`...`) does not play nicely with Neuron, but we should look into other ways to avoid adding `if/else` for each possible number of dims.

File: vllm/model_executor/layers/rotary_embedding.py [hunk 4]
```diff
@@ -253,6 +253,45 @@ def forward_hpu(
...
+        key = key.view(num_tokens, -1, self.head_size)
+
+        if self.rotary_dim == self.head_size:
+            key = _apply_rotary_emb(key, cos, sin, self.is_neox_style)
+            key = key.reshape(key_shape)
+        else:
+            key_pass = key[..., self.rotary_dim:]
+            key_rot = key[..., :self.rotary_dim]
+            key_rot = _apply_rotary_emb(key_rot, cos, sin, self.is_neox_style)
+            key = torch.cat((key_rot, key_pass), dim=-1).reshape(key_shape)
```
Comment (gnovack) [inline vllm/model_executor/layers/rotary_embedding.py:292]: This will need some changes to handle the case where `self.rotary_dim != self.head_size`.

File: vllm/model_executor/layers/activation.py (hunk 3 above)
Comment (aarondou) [reply vllm/model_executor/layers/activation.py:103]: something like the following?
    d = x.shape[-1] // 2
    x_reshaped = x.view(-1, x.shape[-1])
    s = x_reshaped[:, :d] * F.sigmoid(x_reshaped[:, :d])
    result = s * x_reshaped[:, d:]
    return result.view(*x.shape[:-1], d)

File: vllm/model_executor/layers/rotary_embedding.py (hunk 4 above)
Comment (liangfu) [reply vllm/model_executor/layers/rotary_embedding.py]: added handling functionality, along with test cases for checking correctness.

File: vllm/model_executor/layers/activation.py (hunk 3 above)
Comment (liangfu) [reply vllm/model_executor/layers/activation.py]: thanks to @aarondou 's idea, we're able to avoid the if-else condition

File: tests/neuron/test_activation.py (hunk 1 above)
Comment (liangfu) [reply tests/neuron/test_activation.py]: replaced with layer.forward_native

File: tests/neuron/test_rotary_embedding.py (hunk 2 above)
Comment (liangfu) [reply tests/neuron/test_rotary_embedding.py:14]: added related test case, along with function implementation.

**Review by gnovack — APPROVED** [review]
Thanks @liangfu! lgtm
